Improve seek in ogg/opus files

robho opened the pull request:
The file parser had problems seeking in files with "large" ogg pages.

RFC 3533 says:
  Pages are of variable size, usually 4-8 kB, maximum 65307 bytes.

If ogg pages were larger than 9 kB the parser would fail. Ogg and Opus files that I've encoded recently seem to use page sizes ~10-12 kB.

michaelherger commented:
Thanks a lot for your patch! Would you know whether this patch would address any issue reported on forums.slimdevices.com?

While the code looks good with my limited understanding of C, I'm not familiar with either that code or opus as a codec... Would be great if we had a few users who could validate the fix.

@bpa-code @ralph-irving - what's your thought on this?

ralph-irving commented:
Thanks @robho  -
I took a quick look at the changes and it's much cleaner and clearer than the original code.  There have been a lot that's changed since opus support was added to audio scan 5 years ago.
I've applied the PR here and will report back.

ralph-irving commented:
For compatibility with older gcc compilers we'll need to move the loop variable declaration outside the for/next loop.
```
--- src/ogg.c.orig      2023-04-17 10:26:37.085340619 -0400
+++ src/ogg.c   2023-04-17 10:29:40.479712904 -0400
@@ -593,7 +593,9 @@
     }

     unsigned char *bptr = buffer_ptr(&buf);
-    for (unsigned int buf_size = buffer_len(&buf); ; ++bptr, --buf_size) {
+    unsigned int buf_size;
+
+    for (buf_size = buffer_len(&buf); ; ++bptr, --buf_size) {
       if (buf_size < 4) {
        // no page start found!?
        break;
```

I also tried to build for win32 but unfortunately, there are a plethora of errors.  I've attached a log of the failures however due to the cascading errors it gave up after 100.  Getting this to build/work with the VS2010 compiler will require more effort.  Most of the warning are not related to the changes in the PR and have been there for ages, they can be ignored.
The most common failure is declaring variables within scope instead of only at the beginning of a function.  Even my change above is invalid for VS2010! The 2nd is using variable types unknown to VS2010.
[win32buildfailurelog.txt](https://github.com/LMS-Community/Audio-Scan/files/11252253/win32buildfailurelog.txt)

robho commented:
> Would you know whether this patch would address any issue reported on forums.slimdevices.com?

I don't think so. I haven't seen any reports about problems with seek in ogg/opus and I haven't mentioned the fix in the forums either.

I think most people don't notice the problem because:
* I guess most people don't seek in songs
* Most people don't use opus (or ogg (or their ogg files are differently encoded than mine))

> Would be great if we had a few users who could validate the fix.

Yes, the more testing, the better :-)

I did some more testing just now and I see some problem seeking in ogg files on a Squeezebox Radio (squeezelite and squeezelite-esp32 are ok). This radio is running the community firmware. I would appreciate help with testing on other devices and also with the original radio firmware.

I've also seen problems seeking with SB Player for Android but thought that was a player problem since sometimes there are problems seeking in mp3 files there as well.

I'll take a closer look at these problems and see if they are "new" problems and what the cause could be.

> For compatibility with older gcc compilers we'll need to move the loop variable declaration outside the for/next loop.

I'll fix this. I think I also introduced new VS2010 problems by moving variables from the start of a function/scope closer to the place where they're used. I'll revert these changes so that I at least don't make things worse for VS2010. Then maybe we could make an effort later to fix all existing VS2010 problems if we want to support VS2010. Does that sound ok?

ralph-irving commented:
VS2010 is needed build audio scan for the win32 platform.  You'll find that requirement mentioned in README.win32.
Refactoring the PR to only include the changes necessary to fix the opus scan issue would be appreciated and try to use the C89 syntax standard.

robho commented:
I hope the new version I pushed is more VS2010 friendly. I have moved variable declarations to the beginning of scope. If you send VS2010 compile output from ogg.c and opus.c I can make additional fixes.

To help in the future I added -Wdeclaration-after-statement to CCFLAGS (non win32 builds only) so that we get warnings when variable declarations are not in the beginning of the scope. Ok? (I tried -std=c89 but the code isn't compliant (C++ comments for example))

---

Regarding the problems I saw with seeking on SB Player for Android and Squeezebox Radio.. It looks like the problem is that seek fails if I use the material UI on an Android device. I have not determined why it fails, it's strange. Anyway, I've successfully tested the code in this PR on squeezelite, squeezelite-esp32, Squeezebox Radio (community firmware) and SB Player for Android. I've tested both ogg and opus files.

ralph-irving commented:
Great job!  The updated PR builds with VS2010. Thank you.
I won't have time to test the updates on windows until at least tomorrow.

[audio-scan-1.07b3-win32.zip](https://github.com/LMS-Community/Audio-Scan/files/11304316/audio-scan-1.07b3-win32.zip)

robho commented:
Thanks for testing! Take your time, there's no hurry :-)

michaelherger commented on Changes:
```diff
@@ -2,6 +2,9 @@ Revision history for Audio::Scan
 
 Note: Bug numbers refer to bugs at http://bugs.slimdevices.com
 
+1.07	unreleased
```
The version isn't changed in the corresponding `Scan.pm`. Shouldn't this be part of this PR?

robho replied:
Done

michaelherger commented:
Compiled successfully for macOS/Apple Silicon. Running just fine.

ralph-irving commented:
I've been running the changes on windows all week without issue.  Specifically jumping around in opus files with squeezeplay and the web interface.  I only have 3 albums in opus format, so I'm done listening to them now, for a while.
I thought we should review PR#3 and include it in the 1.07 release as well, if accepted.

michaelherger commented:
@ralph-irving feel free to merge this PR. You are part of the team on this repository 😉.
